src/constraints/util/urlValidators.ts

```diff
@@ -0,0 +1,36 @@
+import { ConstraintError } from '../../lib/errors/ConstraintError';
+import { combinedErrorFn, ErrorFn } from './common/combinedResultFn';
+
+export type StringProtocol = `${string}:`;
+
+export type StringDomain = `${string}.${string}`;
+
+export interface UrlOptions {
+	allowedProtocols?: StringProtocol[];
+	allowedDomains?: StringDomain[];
+}
+
+export function createUrlValidators(options?: UrlOptions) {
+	const fns: ErrorFn<[input: string, url: URL], ConstraintError<string>>[] = [];
+
+	if (options?.allowedProtocols?.length) fns.push(allowedProtocolsFn(options.allowedProtocols));
+	if (options?.allowedDomains?.length) fns.push(allowedDomainsFn(options.allowedDomains));
+
+	return combinedErrorFn(...fns);
+}
+
+function allowedProtocolsFn(allowedProtocols: StringProtocol[]) {
+	const oneOf = allowedProtocols.join(', ');
+	return (input: string, url: URL) =>
+		allowedProtocols.includes(url.protocol as StringProtocol)
+			? null
+			: new ConstraintError('s.string.url', 'Invalid URL protocol', input, `expected ${url.protocol} to be one of: ${oneOf}`);
+}
+
+function allowedDomainsFn(allowedDomains: StringDomain[]) {
+	const oneOf = allowedDomains.join(', ');
```

Comment:
Can you move this in the error definition instead, as it's a useless join call for valid domains

Reply:
... what

Reply:
You take this join call
You remove it from here
You move it in the `new Error(` part where its used
You profit

Reply:
Mate, that's not executed in valid domains.

Reply:
Useless execution still :)